### ISSUE TITLE
TaskShims: ignore

### ISSUE THREAD

**Issue by bartelink** [fsprojects/FSharp.Control.TaskSeq#140]
Replaces https://github.com/fsprojects/FSharp.Control.TaskSeq/issues/128. TaskShims top level issue: https://github.com/fsprojects/FSharp.Control.TaskSeq/issues/139

`Async.Ignore` has always been ugly and undiscoverable. While I tend to make ignoring explicit by using `let! _ = <async stuff I want to ignore result of>`, it's commonly the last expression in a function, and having to do `let! _ = <thing I'm wrapping> in ()` is too much.

It is proposed that the `module` associated with any given builder should by convention have an `ignore` function that correctly observes completion of the work, dropping the result, but propagating exceptions, if any
 
Signatures not presently in `FSharp.Core`:
```
module Async =
    let inline ignore (a : Async<'t>) = Async.Ignore
module Task =
    let inline ignore (t : Task<'t>) = ...
module ValueTask =
    let inline ignore (t : ValueTask<'t>) = ...
```

NOTES:
1. Currently, `TaskSeq` (and [other libs such as IcedTasks](https://github.com/fsprojects/FSharp.Control.TaskSeq/issues/128#issuecomment-1353595377) contain various bespoke implementations
2. While this could live in a TaskEx/TaskShims lib, it could be argued that `FSharp.Core` is the obvious home for them. (However that would raise the issue of whether they need to go into the 6.x release line in order to align with minimal dependencies for various common libraries)

**Comment by bartelink** [comment]
Thanks.
1. Fair enough NOTE1
2. Great. For avoidance of doubt, that rules out `Task.ofAsync`
3. Sounds like a taking an `Async.map` + `Task.map` proposal around the houses could flesh out a lot of the pros/cons, leaving filling out the rest of the matrix a more constrained ask. NOTE2
----  
NOTE1 but I HATE the Async.Ignore wart; thing is I can add an `inline` one-liner in my personal helpers. Its very easy for these warts to drive people to the conclusion that the answer is to just use `Task` for all the things and treat `Async` as some unfortunate historical path. i.e. if I can do `Task.ignore` and `ValueTask.ignore` but Async is not in alignment.)

NOTE2 But: If it's in the box, it becomes blessed. It will inevitably lead to clear 3 line `async` or `task` blocks becoming more terse piped equivalents, for better or for worse (it's clearly subjective)

I personally think we're better off without it, but definitely not a hill I want to die in, especially if having some agreement means we have consistent Task/Async/ValueTask helper names going forward, and a pattern for FsToolkit, IcedTasks and/or other computation expressions to follow.

**Comment by T-Gro** [comment]
> do low value and/or arguably duplicative shims and sugar really belong in FSharp.Core, or would a dedicated lib in fsproject make more sense?

We should only do it if we believe an addition is the better one of the duplicates.
If it is - we keep the "worse" for source+binary back compat and apply attributes like Obsolete or EditorBrowsable.Never.
If it is not - do not do it (even though it exists at some places of fslib now).

> having map or not

If there is a common need, we should provide a default.
That makes even more sense with agentic coding and reducing the chances of supply chain attacks.

> who are the realistic maintainers of such an API set?

Technically, the RFC process + eventually maintainers of dotnet/fsharp responsible for meeting support+security commitments.
I am more then happy to explore improvements to the process of API co-ownership for FSharp.Core that would build on top of that (i.e. not replace, but provide an additional guard via empowered individuals).

Without losing the openness and asynchrony of the current process - it is what allows people from across the world to contribute to F#'s evolution.

@bartelink :
Would you be interested in driving a "Async+Task+ValueTask" library improvements collection - an umbrella suggestion with the goal to present API additions to the related modules for .NET 11 ? (open e.g. until end of May. I/we could then take it for RFC+impl, so that it becomes available for release candidate versions in time?)

**Comment by bartelink** [comment]
@T-Gro Yes, I'd be interested and should have time. It would certainly help for the noise issues I added here to move out (and this lib would also benefit from being able to drop some polyfills it hosts).

Any particular things to rule in/out or add to this laundry list of things I'm aware of in this space?
1. I assume such an effort would consider the [AsyncEx issues here](https://github.com/fsprojects/FSharp.Control.TaskSeq/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22topic%3A%20task-ex%22)
2. perhaps the library aspects of longstanding issues such as:
    - `AwaitTask` should not egregiously wrap in `AggregateException` aka [`Async.Await`](https://github.com/fsharp/fslang-suggestions/issues/840)
      - inc perhaps [related `AggregateException` anomalies in adjacent Async.* methods](https://github.com/fsharp/fslang-suggestions/issues/840#issuecomment-1935540325)
    - `Async.Await` should honor cancellation of ambient CT: https://github.com/dotnet/fsharp/issues/2127
    - `Async.Await` and/or tooling should encourage/warn about failure to propagate CT to inner `Task`: https://github.com/fsharp/fslang-suggestions/issues/1284
3. naming consistency policy - do we provide *.Ignore, *.ignore (and hide Async.Ignore), or go down the road of Async.Ignore + Task.ignore +ValueTask.ignore
4. consistency of available methods/functions in Async vs Task vs ValueTask (e.g. an appropriate `ignore` mechanism)
5. hiding / de-emphasizing policy i.e. Warns and/or other tooling hooks re `Async.AwaitTask` deprecation in favor of `Async.Await` etc (potentially via `Obsolete` or `EditorBrowsable.Never`)
6. Longstanding debatables such as whether `Async.map` should go in the box (inc Task.map and ValueTask.map) 

What are the next steps? writing an umbrella/epic in dotnet/fsharp?

**Comment by T-Gro** [comment]
I am happy to read that! 👍  Thanks for making F# better.

Ideally, I would love to see this as a new fslang-suggestion (since that repo is the place where people go to watch out for F# evolution, and provide feedback), an umbrella for all things "Async/Task/ValueTask API additions to FSharp.Core".

From a product perspective, I would like those improvements to be released together - hence a single umbrella (with links to any preexisting suggestions) instead of many issues in isolation.

1,2,4 - yes
3 - consistency + hiding
5. If just naming inconcistent, only hide. If deemed wrong, Obsolete marker with warning (IsError=false) explaining what to use instead and why that is.
6. yes. The signals (high replication rate, being offered from multiple libs..) are there.

**Comment by bartelink** [comment]
@T-Gro I have a WIP skeleton https://gist.github.com/bartelink/a19865472e5751f70210c263655676b4
Would like your review/comments before I post to fslang

High level decisions I'm [implicitly] pushing for:
- stick with individual named functions (don't go down the road of a `module` with camelCase methods masquerading as functions (i.e. having to have `ofTask` and `ofTaskUnit` vs having the possibility/confusion of two (or more) overloads of a camelCase named thing that fits the general function named style
- The two new `Async.Awaits` go as overloaded methods on `type Async`
  - **have to decide whether `Async.Await` and `Async.AwaitTask` remain visible side by side (or use Browsable.Never for AwaitTask)**
- **`module Async` provides forwarding functions (`ofTask`, `ofTaskUnit`) per overload.**
  - _OR remove `Async.ofTask` and `Async.ofTaskUnit` in favor of making people pick from `Async.Await` vs `Async.AwaitTask` for better alignment with established API, and an easier way to explain and/or tool migration?_
    - i.e. `AwaitTask` becomes `Await` vs looking like a style change if those became `ofTask`
    - however that does break symmetry - Task and ValueTask *will* have various `ofTask` converters while `Async` will not etc (but then again there are no semantic implications and/or considerations as there are for `AwaitTask` vs `Await` to the same degree)
- no functions that start Asyncs running (with or without CT flowing) as part of this particular proposal

Potential followups (let me know if any should be done soon; my sense is that it's best to leave it until this proposal has garnered feedback and been adjusted to take account of same?)
1. pipeable start child flowing CT helper: `val startImmediateAsTask: CT → Async<'T> → Task<'T>`. NOTE this should be named along with the next one (`call`/`exec`)
1. flowing the Ambient CT into a child Task invocation helper yielding an `Async<'T>`: `val call: (CT -> Task<'T>: Async<'T>` (or `Async.run` / `Async.runTask` to allude to `Task.Run` semantics?). Semantics approx `(fun f -> Async.CancellationToken >> f >> Async.Await)`. Potential alternate name `Async.exec`. Note this would replace/address https://github.com/dotnet/fsharp/issues/2127
1. pit of success `Async.Parallel`: `val parallelLimit: int -> seq<Async<'T>> -> Async<'T[]>`
1. pit of success `Async.Parallel<unit> -> Async<unit>`: `val parallelDoLimit: int -> seq<Async<unit>> -> Async<unit>`

**Comment by bartelink** [comment]
@T-Gro I posted https://github.com/fsharp/fslang-suggestions/issues/1466 in the interests of getting it out there and will now close this. Decisions I made re the above include:
- No `Obsolete`/`BrowserVisible` attributes
- stick with `module` for this proposal
- Don't integrate with `Async.Await`/`Async.AwaitTask`
  - It's orthogonal
  - most logical way to manage migration is for them to live adjacent to each other and be PascalCase
  - there are no functions that start Asyncs in the proposal (due to CT propagation concerns) so less need
  - the fact it's a module would mean the cumbersome `ofTaskUnit` vs `ofTask` naming does not come into the picture

**Comment by bartelink** [comment]
Closing as this is now covered by https://github.com/fsharp/fslang-suggestions/issues/1466

**Comment by T-Gro** [comment]
I agree with the decisions here 👍 .

- `Async.Await` becoming the single recommended API to use makes sense. Using Browsable.Never for existing `AwaitTask`
- The `of**` pattern is common also in collections, I think the forwarding makes sense here. It enables easier discovery by following that pattern.
- I think the followups make sense and I have been thinking this might be a single umbrella suggestions for all things "Async/Task API improvements". 
- I would prefer the `start` naming over `exec`/`call`. But in general agree for providing functions (the PIT of success category) that require arguments for those to work successfully 👍

**Comment by bartelink** [comment]
@T-Gro 
>Async.Await becoming the single recommended API to use makes sense. Using Browsable.Never for existing AwaitTask

Proposal rules out Browsable.Never. Unless you rethink I can adjust it to propose it (I'm thinking of people typing `Async.Await` and not getting an `AwaitTask` autocompletion for the first time in 15 years and wondering what's going on - I know the xmldoc could explain "if you really need strict backcompat, `AwaitTask` still exists but type it manually to confirm" ? Is there prior art for hiding this to this degree?

>The of** pattern is common also in collections, I think the forwarding makes sense here. It enables easier discovery by following that pattern

in the final proposal I did not put in `ofTask` or `ofValueTask` etc as:
- while there's an obvious correct decision (no AggregateException, honor Async cancellation), burying that in an `of` is a bit debatable
- because you'd have to have functions for plain Task, Task<T>, ValueTask, ValueTask<T>, each one needs a name (assuming you don't go down the use a `type` so you can overload functions hack road). I don't like `ofTaskUnit`, `ofUnitTask` so unless there's a good idea I don;t see how it would work?
- there's prior art for having a pascal cased `Async.AwaitTask` and/or `Async.AwaitTaskCorrect` in lots of code already?

Can you tell me the specific names/signatures you'd stick in? I can definitely see the pros but I need to be convinced overall...

> I think the followups make sense and I have been thinking this might be a single umbrella suggestions for all things "Async/Task API improvements".

OK I'll remove some of the cross references/commentary from the Async/Task/ValueTask suggestion OP and the https://github.com/fsharp/fslang-suggestions/issues/1466#issuecomment-4363497441 comment below it and post a new suggestion that lays out the overall set of features being shortlisted:
- Add Async.Await i.e. https://github.com/dotnet/fsharp/issues/840 (inc honor cancellation covering https://github.com/dotnet/fsharp/issues/2127)
- Add Async/Task/ValueTask helpers
- Add Async.start* APIs covering territory of https://github.com/fsharp/fslang-suggestions/issues/1284

  >I would prefer the start naming over exec/call. But in general agree for providing functions (the PIT of success category) that require arguments for those to work successfully 👍

  OK, `start` it is ;) Will followup in the umbrella issue and/or 

Any thoughts or preference re the `empty` I added to the proposal last minute? Should it stay, be changed to something like `zero` or be removed for now so it can have a longer gestation and/or a debate as it's own ticket after the dust has settled?

**Comment by T-Gro** [comment]
1. EditorBrowsable.Never
    - I am not aware of any precedents. We can use strong remarks in public docs of both (Await as well as AwaitTasks) instead.
    - Alternative might be a full-blown ObsoleteAttribute, but that feels too strong in this case.
2. ofT..
    - I like what is in the fslang suggestion now 👍 
3. empty
    - I am not sure I follow how this fits generic arguments and the `Task` / `Task<T>` schizm. 
    - Or did you mean this just for `unit` (or non-generic variants), and not as an overall API for `Completed<T>(value)` functionaity ?

**Comment by bartelink** [comment]
@T-Gro 
>Or did you mean this just for unit (or non-generic variants), and not as an overall API for Completed<T>(value) functionaity ?

This is all about a thing *not* returning a value; `singleton` is the API for a Completed state Task<'T> or ValueTask<'T>

The current proposed impls are:
- `Async.empty: Async<Unit>`: same as `async.Zero()` and/or an `async {` without a `return` I guess)
- `Task.empty`: non-generic Task, same as Task.CompletedTask - no result, just Completed synchronously
- `ValueTask.empty`: again, non generic. Note netstandard2.1 does not provide a ValueTask.CompletedTask, which is one reason that the standard helpers in TaskSeq has it

So in other words, no generic vs non generic variants (and/or need to call them `ValueTask.empty` vs `ValueTask.emptyUnit` or something to distinguish between the ValueTask<unit> and ValueTask variants. But maybe I'm missing something (I've never wanted/needed a pre-built/shared `Task<unit>` AFAICR)

The main reason I would think it useful is that if you have a `singleton`, a `map` and a `bind`, any pipeline based Task work will not involve any `open System.Threading.Tasks`, so if you're trying to pass in a zero/placeholder task, having to do `task { }` or `System.Threading.Tasks.CompletedTask` breaks the pattern vs being able to do `Task.empty`.

But, of course, it's also not very discoverable and is a bit ad hoc:
- nobody really does `async.Zero()` in code, and not many more will do/discover `Async.empty` vs `async { () }` (or `async { }` as of F# 8 IIRC?)
- anyone coming from C# will probably just type Task.CompletedTask and auto-open the namespace; a `Task.empty` would not be something that would enter your head

I don't think `completed`, `completedTask`, `zero` or any other thing that enters my head is better... I guess for containers that can have content (and the absence thereof), the `empty` vs `isEmpty` functions make sense.

I think I've talked myself out of it already - unless you have a better name or think there's something there? (even then, perhaps I should just write an individual suggestion so people can refer to a short canonical writeup on the concept?)

**Comment by bartelink** [comment]
@T-Gro edited in notes based on the above into fslang suggestion. **For now I've left `empty` in, corrected some signatures, and added a more succinct version of the above - but let me know if you think it's best to remove it - we should take a final stance before spreading the suggestion more widely...**.

**Comment by bartelink** [comment]
@T-Gro  re I've posted a 70% complete proposal re https://github.com/fsprojects/FSharp.Control.TaskSeq/issues/140#issuecomment-4390679108 at https://github.com/fsharp/fslang-suggestions/issues/1467 

I'll start to expand it further in a few hours / tomorrow am

After that, it and https://github.com/fsharp/fslang-suggestions/issues/1466 should be publicized more widely

Comments from you or anyone else below aimed at making the proposal as complete and non controversial from the off appreciated👇

**Comment by bartelink** [comment]
@T-Gro I think it's at a checkpoint now. I've got
- The standard functions: https://github.com/fsharp/fslang-suggestions/issues/1466
- The interop helpers: https://github.com/fsharp/fslang-suggestions/issues/1467

The `sequential`, `parallelLimit` and `parallelDoLimit` functions have been proven in a closed source codebase, in addition to internally in Equinox and Propulsion, though I'm not aware of them being in any other libraries

Happy to take comments either below, or on https://github.com/fsharp/fslang-suggestions/issues/1467

@TheAngryByrd Given you own the a few of the suggestions it's built on, I'd definitely be interested in any early feedback before this gets shared more widely - I'm not wed to the names (my current ones are derived from https://github.com/jet/propulsion/blob/master/src/Propulsion/Internal.fs#L125-L168 and looking at FsToolkit)

**Comment by bartelink** [comment]
@TheAngryByrd should probably discuss this on https://github.com/fsharp/fslang-suggestions/issues/685 itself but I can't figure out a way to include it in the overall featureset laid out in https://github.com/fsharp/fslang-suggestions/issues/1467

```fsharp
module Async =
  let withCancellation (ct:CancellationToken) (a:Async<'a>) : Async<'a> = async {
    let! ct2 = Async.CancellationToken
    use cts = CancellationTokenSource.CreateLinkedTokenSource (ct, ct2)
    let tcs = new TaskCompletionSource<'a>()
    use _reg = cts.Token.Register (fun () -> tcs.TrySetCanceled() |> ignore)
    let a = async {
      try
        let! a = a
        tcs.TrySetResult a |> ignore
      with ex ->
        tcs.TrySetException ex |> ignore }
    Async.Start (a, cts.Token)
    return! tcs.Task |> Async.AwaitTask }
  
    /// Async.Start with timeout in seconds
    let startWithTimeout (timeoutSecs:int) (computation:Async<unit>) =
      let c = new System.Threading.CancellationTokenSource(timeoutSecs*1000)
      Async.Start(computation, cancellationToken = c.Token)
```

It seems to me that a generic way to link in a cancellation token into a child task invocation might be to extend Async.StartChild to accept a CT as well as a timeout (presumably the timeout is implemented via a linked CT?), see https://fsharp.github.io/fsharp-core-docs/reference/fsharp-control-fsharpasync.html#StartChild

Or am I missing the point entirely? Is this for top level execution, or for managing some child of the overall computation tree?

----

In that issue, I also posted a different internal helper from Propulsion:-
```fsharp
module Task =

    /// Runs an inner task with a dedicated Linked Token Source. Cancels via the ct upon completion, before Disposing the LCTS
    let inline runWithCancellation (ct: CancellationToken) ([<InlineIfLambda>]f: CancellationToken -> Task) = task {
        use cts = System.Threading.CancellationTokenSource.CreateLinkedTokenSource(ct) // https://stackoverflow.com/questions/6960520/when-to-dispose-cancellationtokensource
        try do! f cts.Token
        finally cts.Cancel() }
```
I use that helper to do essentially achieve the same thing (something along the lines of a `Task.StartChild`), see examples: https://github.com/search?q=repo%3Ajet%2Fpropulsion%20runWithCancellation%20&type=code

But again, I'm not convinced it's a proper generalizable pattern as such so have not included anything like that.

If you can give me a rough idea of what one'd use it for, I'd definitely be interested to include it if I can work it into the design - even if we don't include it in the API set this time around, the exercise of seeing how it should be named to align with the others would be extremely useful

**Comment by TheAngryByrd** [comment]
> It seems to me that a generic way to link in a cancellation token into a child task invocation might be to extend Async.StartChild to accept a CT as well as a timeout (presumably the timeout is implemented via a linked CT?), see https://fsharp.github.io/fsharp-core-docs/reference/fsharp-control-fsharpasync.html#StartChild

> Or am I missing the point entirely? Is this for top level execution, or for managing some child of the overall computation tree?

The original use case was using the [RequestAborted](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpcontext.requestaborted?view=aspnetcore-10.0) CancellationToken from a http handler. `Async.withCancellation` It's more closely aligned to `Async.StartAsTask` but not necessarily leaving `Async` return type. 

But at this point with Async/Task world, I'd forego Async and just use `Async.StartAsAsync`.  There' rarely a need for Async anymore in my work outside MailboxProcessors.

**Comment by bartelink** [comment]
@TheAngryByrd 

> `StartAsAsync`

I take it you mean Async.StartImmediateAsTask(computation = ..., cancellationToken = ct)

Based on the stuff I have in https://github.com/fsharp/fslang-suggestions/issues/1467 you'd do use one of:

```fsharp
let f args = async { ... }

member _.Action ct = Task.runAsync (f args) ct
```

OR

```fsharp
member _.Action ct =
    Task.startAsync ct <| async {
        return! f args
    }
```
OR

```fsharp
member _.Action ct =
    async {
        return! f args
    } |> Task.startAsync ct 
```

IME unless you have IcedTasks/CancellableTask/async2 in your stack, `async` can still make sense for quite a few layers as you get cancellation token propagation without it being in your face, but ASP.NET and/or a C# top layer can speak Tasks and explicit cancellation token wiring?

What's in the box is important or people without the full array of tools mapped out in their heads will say "hm, well in C# and ASP.NET I pass a `CancellationToken` everywhere and end up just using bald `task {` and/or punting on correct cancellation token propagation like the average C# impl?

**Comment by bartelink** [comment]
@T-Gro re https://github.com/fsharp/fslang-suggestions/issues/1467#issuecomment-4430846128

> I assume its a typo and `Task.parallelDoLimit` should be having `Task`-based inputs and output, not `async` anywhere?
> If it is not a typo, I agree with Jimmy - regular actions (i.e. not `awaiters` or explicit `ofXXX` converter) should not flip.

Yes, sorry, once again wasn't paying attention (can you remove the comment to make it less noisy / protect my pride though please!)

Also I unmangled the Task.runAsync vs Task.startAsync in 1467.

The most commonly useful signature from a piping perspective is the `startAsync` one (i.e. `startAsync: CT -> computation -> Task`) even if logically it's `runAsync: computation -> (CT -> Task)` 

But having two copies of the same thing under two names with just flipped arg order feels wrong.

If I am only allowed one, I think I'd settle for `module Task = let startAsync: CT -> computation -> Task = Async.StartImmedateAsTask(computation, cancellationToken = ct)` ?

**Comment by bartelink** [comment]
cc @gusty I'm thinking you've been around the houses on many of these things

If you have the time would appreciate some comments here re https://github.com/fsharp/fslang-suggestions/issues/1466 and https://github.com/fsharp/fslang-suggestions/issues/1467 - especially if there's stuff you really don't like, consider superfluous, or a minor tweak would improve it. Ditto naming improvements before we spread it wider (maybe via @sergey-tihon ?) would be more than welcome.

Is there anyone else that should be CC'd on this for a first round of getting rid of obvious cruft or extremely commonly helpful/needed small functions? @eiriktsarpalis

**Comment by T-Gro** [comment]
I removed that comment 👍 , thanks for fixing it.

Once you are ready with final touches on those two suggestions, we might invite people via Discord - this is for sure the most active F# social and people will come with opinions :).

re: startAsync vs runAsync - 
I would pick startAsync as well (it's taking a computation and starting it into a hot task)

**Comment by bartelink** [comment]
@T-Gro Thanks

@T-Gro I don't have any further things I want to change in either suggestion at this point, but just want to finalize re startAsync/runAsync as for me your comments are inconclusive re the following:

```fsharp
type Async with
     // https://github.com/fsharp/fslang-suggestions/issues/1042
     // NOTE name is so it appears in completion list near Async.RunSynchronously
     // Still open to suggestions to intuitively convey that this is RunSynchronouslyCorrect though - see #1042
     member _.RunSynchronouslyImmediate(computation: Async<'T>): unit =
         // runs on this thread, without implicit hop to thread pool, so has a correct stack trace
     // NOTE name has Task in it to a) allude to Task.Run b) align with [Value]Task.runAsync
     member _.RunTask(createTask: CancellationToken -> Task): Async<unit> = ...
     ... and further overloads of RunTask ...
module Task =
     // NOTE name aligns with Async.runTask, forces flowing of cancellation
     let runAsync (computation: Async<'T>): CancellationToken -> Task<'T> = // Async.StartImmediateAsTask...

     // TOCONSIDER more natural for use with piping - as per runAsync but with the arguments flipped

     //  member _.Action1(arg, ct): Task<'T> =
     //       async { do! x arg; return y }
     //       |> Task.startAsync ct
     //  member _.Action2(arg, ct): Task<'T> = Task.startAsync ct <| async {
     //       do! x arg
     //       return y }
     //  member _.Action3(arg, ct): Task<'T> =
     //      Task.startAsync ct <| async {
     //           do! x arg
     //           return y }
     //let startAsync (ct: CancellationToken) (computation: Async<'T>): Task<'T> = runAsync computation ct
```

IME it's very clear from lots of usage that `startAsync` is the more useful signature (in terms of arg order, even if the CT is logically the last part that should be being supplied from a partial application perspective).

The questions to be decided are:
- is the fact the arg order is slightly illogical a concern (probably moot if nobody is going to use a technically more logical one)
- should both be present (`Task.runAsync` is the logical inverse of `Async.runTask`)
- if only one is present, which one, and what name
- assuming its one function and the name is based off `start`, should it be `Task.start`, `Task.startImmediate`, `Task.startAsyncImmediate` or something else given there's an `Async.RunSynchronouslyImmediate` being added at the same time ?

It'd be great to have your thoughts on this so the first cut isnt leaving lots of stuff open (its not a problem if people come with other opinions; it's more about having something concrete so any improvement/alteration suggestions can be equally concrete)

**Comment by TheAngryByrd** [comment]
> IME unless you have IcedTasks/CancellableTask/async2 in your stack, async can still make sense for quite a few layers as you get cancellation token propagation without it being in your face, but ASP.NET and/or a C# top layer can speak Tasks and explicit cancellation token wiring?

> What's in the box is important or people without the full array of tools mapped out in their heads will say "hm, well in C# and ASP.NET I pass a CancellationToken everywhere and end up just using bald task { and/or punting on correct cancellation token propagation like the average C# impl?

I'm not arguing it wouldn't be useful, there are plenty of times you have an outside `CancellationToken` and an `Async<_>` and you want to fuse the two. ASP.NET is just the time that shows up regularly with this scenario. Other scenarios like in [fsac](https://github.com/ionide/FsAutoComplete/blob/800da164952f21ed720de172a0ddb811af66e956/src/FsAutoComplete/LspServers/AdaptiveServerState.fs#L483-L485) also have this problem because a cancellation can come from an LSP client.

I'm just saying that I don't have this scenario often anymore as I tend to use IcedTasks when possible.

**Comment by gusty** [comment]
Thanks @bartelink  I added my thoughts to the discussion.

Regarding the replacements of `Async.StartImmediateAsTask` isn't it worth a separate issue for that (and revisiting the default cancellation token stuff), as [you suggested](https://github.com/fsharp/fslang-suggestions/issues/840#issuecomment-1940891330) long time ago?

**Comment by bartelink** [comment]
> Other scenarios like in [fsac](https://github.com/ionide/FsAutoComplete/blob/800da164952f21ed720de172a0ddb811af66e956/src/FsAutoComplete/LspServers/AdaptiveServerState.fs#L483-L485) also have this problem because a cancellation can come from an LSP client.

@TheAngryByrd For the needs of the fsac usage above, assuming what I think T-Gro is asking for, AIUI, this could be written as:

```fsharp
UnusedOpens.getUnusedOpens (tyRes.GetCheckResults, getSourceLine)
|> Task.startAsync progress.CancellationToken
|> Async.Await // assuming asyncEx does not Bind to `Task<'T>`
```
Of course I may be entirely missing the point ;)

I'm really just trying to bottom out on whether the Async.withCancellation suggestion can be covered or 90% covered by some combination of smaller helpers - i.e. if there is a real need an impl involving `TaskCompletionSource` gymnastics and it's remotely common, putting it in the box is on the table. But if we're providing a `Task.startAsync` that has a non-optional `CancellationToken` and the only awkwardness is whether an `Async.Await` is required based on the specific bindings available at the call site then I'd be seeking to close that suggestion with a workaround of "this can be done succinctly via `Task.startAsync ct |> Async.Await`"

**Comment by bartelink** [comment]
Hi @gusty and thanks for responding.

Thanks so much for the detailed response on 1466 - will respond over there regarding those.

I'm a bit more confused re your other points so apologies in advance for the wall of text below ;)

>Regarding the replacements of Async.StartImmediateAsTask isn't it worth a separate issue for that (and revisiting the default cancellation token stuff)

I was aware of the discussion you linked (I've traversed all these tickets including all discussions in depth in recent times). In that case I was suggesting that there should be a separate tracking item wrt stuff that brings the `Async.DefaultCancellationToken` into play. That's various `type Async` `member`s that have a `?cancellationToken` argument. The suggestions here are specifically about functions  on `module Async` and/or `module Task`, and each of these have a non-optional `CancellationToken` (the caller is forced to explicitly use `CancellationToken.None` or `Async.DefaultCancellationToken` or something else as appropriate)

Within this context, https://github.com/fsharp/fslang-suggestions/issues/1467 is intended to cover extensions to `type Async`, `module Async` and `module Task` wrt starting Asyncs, viz:
- `Async.RunSynchronouslyImmediate` (has a specific ticket https://github.com/fsharp/fslang-suggestions/issues/1042 but the names should align - if `Immediate` is the term for 'no thread pool / SynchronizationContext induced hops')
- `Async.RunTask` (has a specific ticket https://github.com/fsharp/fslang-suggestions/issues/1284)
- `Task.startAsync` - one liner that forces a CT and calls `Async.StartImmediateAsTask`

You are correct that there is no specific ticket re `[Value]Task.startAsync`, but I think the questions in https://github.com/fsprojects/FSharp.Control.TaskSeq/issues/140#issuecomment-4442238003 can likely be resolved here before we start a long discussion

In other words, I'd love to have an outline agreement on the full set of signatures presented in 1467 before we consider to mint a Task.startAsync ticket

Similarly the parallelLimit/parallelDoLimit - if we can rule doing something like that in or out at this point, then a separate ticket can be done for completeness.

> Regarding the replacements of `Async.StartImmediateAsTask`

I'm not proposing to replace/change anything on `type Async` as such (and I'm trying to keep things tractable and concrete by limiting myself to adding things to the library vs talking about adding warnings and/or having changes as such) . I did consider going down that road but as detailed in Appendix A of https://github.com/fsharp/fslang-suggestions/issues/1466 IMO:
- Async.AwaitTask cant/shouldn't be made BrowserVisible.Never as there's way too much code that's likely dependent on specifics of its quirks
- Async.Await needs to be on the type for overloading/discovery reasons
- I'm specifically suggesting the xmldoc for Async.Await and Async.AwaitTask will be doing most of the lifting (it's not inconceivable that something later suggests migrating off Async.AwaitTask)
- Async.Catch and Async.catch need to coexist even if ideally people switch to a catch, with a `Result` result (which aligns with `Task.catch`
- Async.RunSynchronously, Start, StartAsTask, StartWithContinuations, StartWithContinuationsUsingDispatchInfo, StartImmediateAsTask, StartImmediate all have a `?cancellationToken` that defaults to `Async.DefaultCancellationToken`, which is hard to discover and can be a source of leaks in that each resulting invocation subscribes to the default CT  

So, what are you specifically suggesting we should make a ticket for? How many tickets? Are you suggesting they should be considered for the F# 11 timeframe?

**Comment by gusty** [comment]
Thanks for the clarification, I agree on having them in the same ticket.

**Comment by TheAngryByrd** [comment]
> I'm really just trying to bottom out on whether the Async.withCancellation suggestion can be covered or 90% covered by some combination of smaller helper

It probably can be handled by small helpers. Might be worth considering how this affects Fable. They don't support `task` and would need to shim this as transition from `async -> task -> async`. So having an "async native" approach might be more beneficial there.

**Comment by bartelink** [comment]
> So having an "async native" approach might be more beneficial there.

I'm not feeling sure we have proven there's a common enough pattern/need.

It seems to me we have a variety of ways of starting tasks that admit a CT directly:- Async.RunSynchronously, Start, StartAsTask, StartWithContinuations, StartWithContinuationsUsingDispatchInfo, StartImmediateAsTask, StartImmediate 

While layering in cancellation below the top level might make sense in some cases, it feels like that belongs in with StartChild (which has a timeout option and involves LinkedCancellationTokenSource etc).

In other words
a) I don't feel we have enough of a handle on it to include it in this apiset so won't be attempting to address the need in 1466/7 for F# 11 timeframe
b) I'm not convinced the ticket as it stands is helping matters even in the longer term - I suspect most people that think they want it can achieve their goals via the existing starting functions [and not having to descend into complex TaskCompletionSource juggling]

**Comment by bartelink** [comment]
@T-Gro sorry for the bombardment of atting but wanted to close out on https://github.com/fsprojects/FSharp.Control.TaskSeq/issues/140#issuecomment-4442888110

I'm guessing that removing `Task.runAsync` and having a single `startAsync` with the CT arg first is your preference, but the question of including `Immediate` in the name to:
a) convey the semantics
b) allow other such helpers to sit alongside either in the box or via augmentations from libraries or app preludes

still needs answering from my perspective?

I'm also wondering whether `Async.RunTask` should then more logically be called `Async.RunTaskImmediate` to convey that it will be using StartImmediateAsTask and not doing a thread pool or SynchronizationContext hop ?

cc @TheAngryByrd I'm guessing you might have thoughts on the naming here too
cc @gusty you may have less interest in this one but if you have any thoughts ...
@gusty also I expected you might object to `empty` on similar grounds to `singleton` (= async.Zero / Task.FromResult(()) / ValueTask(())) ? (I'm considering removing it in favor of `result ()` given `Task` becomes `Task<unit>` across the board)

Will update 1466 and 1467 with proceedings of this and https://github.com/fsharp/fslang-suggestions/issues/1466#issuecomment-4443666240 and pop a link into the discord and the slack when I'm done with the editing later and/or tomorrow am

**Comment by gusty** [comment]
> I'm also wondering whether Async.RunTask should then more logically be called Async.RunTaskImmediate to convey that it will be using StartImmediateAsTask and not doing a thread pool or SynchronizationContext hop ?

Definitely yes. `RunTask` alone is a bit vague, given the different options available.
Same goes for including Immediate in the other function name.

**Comment by bartelink** [comment]
Applied naming updates and guesses on changes to https://github.com/fsharp/fslang-suggestions/issues/1467

**Comment by T-Gro** [comment]
I like the changes done.

It still holds that the more technically detailed explanation must sit in API docs, and there should be an unified way of explaining common terms (a shared glossary file perharps, if I decide to land `/// <include file="{share glossary xml}" path="/async-task-glossary/immediate"/>` work which has a draft PR. Forcing use to use a shared glossary across Async/Task/ValueTask would ensure that same english words carry the same meaning across the API set ) :
`start` / `run`
`immediate`
`result`
`catch`


As a basic test, are we able to one-sentence explain the difference between `start` and `run` ?

**Comment by bartelink** [comment]
@T-Gro A very good question

## Glossary

### Reasonably clear terms:
- Synchronously - blocking the current OS thread (not for general use in a tree of computation)
- Immediate - starts on current thread with no hops (which might be bad)
- Child - shares/propagates/links cancellation token; child runs concurrently

### Less clear terms:

`Run`
  - kick off somewhere and give me a handle to it
  - Alludes to Task.Run (all overloads return a Task)
  - IMO Task.Run is a common well known thing for most .NET devs

`Start`
  - used in names of lots of Async things including MailboxProcessor
  - anything Async can be started to run parallel or concurrently depending on lots of factors
  - in general is intended to convey we're not doing stuff inline - we're kicking off parallel work here is the implication

`AsTask`
- Because there was no `Task` in FSharp.Core historically, lots of things that arguably belong on that side of the fence (and would simplify the naming and documentation of `Async.(Start|Run).*`) are mingled in and need disambiguation

## Survey

| Async API | Might ideally be called | returns | notes |
|--|---|---|---|
| `RunSynchronously` | `Async.RunBackgroundSynchronously` | 'T | blocks, thread hop if not thread pool thread |
| `RunSynchronouslyImmediate` (**NEW**) | fine? | 'T | blocks, no thread hop |
| `Start` | `Async.StartBackground` | `unit` (computation must be `Async<unit>`) | thread pool only |
| `StartAsTask` | `Task.startAsyncOpt` | `Task<'T>` | admits `TaskCreationOptions` |
| `StartImmediate` | `Async.StartImmediate` | `unit` (computation must be `Async<unit>`) | starts immediate, then thread pool |
| `StartImmediateAsTask` | `Task.startAsyncImmediate` | `Task<'T>` | starts immediate, then thread pool |
| `StartChild` | `Async.StartChild` | `Async<Async<'T>>` | Flows CT, admits optional timeout |
| `StartChildAsTask` | `Task.startAsyncChildOpt` | `Async<Task<'T>>` | Flows CT, admits `TaskCreationOptions` |
| `StartWithContinuations` | fine? | `unit` (computation must be `Async<unit>`) |
| `RunTaskImmediate` (**NEW**) | `Async.StartTaskImmediate` | `Async<'T>` | `Async.CancellationToken >> Async.StartImmediateAsTask >> Async.Await` |

## Conclusion

1. *Start* is not a useful word anywhere than in `Async.`, where it's primary role is to identify things that instigate Async computations in various ways
2. *Run* is similarly meaningless and is also overloaded (RunSynchronously is actually very similar to Task.Run >> _.Wait() semantically), but people think it runs things inline like the proposed `RunSynchronouslyImmediate` will do [more of].

- In the above `RunTaskImmediate` is the only one using `Run` in the name, and, being `Immediate` is actually not close enough to `Task.Run` semantics so probably `StartTaskImmediate` is a better name. Also `Start` references `StartChildAsTask` / `StartChild` like semantics
- `Run` or `Start` as they stand are not worth/possible to put in a glossary as they are meaningless and/or confusing. If `RunSynchronouslyImmediate` is the only new function with `Run`, there's less need to talk about the difference

Hopefully this triggers someone to see a better naming scheme.

~For now unless there are objections, I'm leaning toward changing the proposed `Async.RunTaskImmediate` to `Async.StartTaskImmediate` to take the start vs run 'debate' off the table.~ I have updated `RunTaskImmediate` to be named `StartTaskImmediate` based on the above
